### ISSUE TITLE
nc_rename_dim() silently produces corrupt .nc4 file 

### ISSUE THREAD

**Issue by czender** [Unidata/netcdf-c#468]
The test file is [here](http://dust.ess.uci.edu/tmp/merra.nc4). This command 

`ncrename -O -d record,mth ~/merra.nc4 ~/foo.nc`

exits without error yet produces an output file unreadable by ncks or ncdump:

```
zender@skyglow:~$ ncrename -O -d record,mth ~/merra.nc4 ~/foo.nc
ncrename: In total renamed 0 attributes, 1 dimension, 0 groups, and 0 variables
zender@skyglow:~$ ncdump ~/foo.nc
HDF5-DIAG: Error detected in HDF5 (1.8.19) thread 0:
  #000: H5L.c line 1181 in H5Literate(): link iteration failed
    major: Symbol table
    minor: Iteration failed
  #001: H5Gint.c line 842 in H5G_iterate(): error iterating over links
    major: Symbol table
    minor: Iteration failed
  #002: H5Gobj.c line 696 in H5G__obj_iterate(): can't iterate over compact links
    major: Symbol table
    minor: Iteration failed
  #003: H5Gcompact.c line 427 in H5G__compact_iterate(): iteration operator failed
    major: Symbol table
    minor: Can't move to next iterator location
  #004: H5Glink.c line 480 in H5G__link_iterate_table(): iteration operator failed
    major: Symbol table
    minor: Can't move to next iterator location
HDF5-DIAG: Error detected in HDF5 (1.8.19) thread 0:
  #000: H5L.c line 1181 in H5Literate(): link iteration failed
    major: Symbol table
    minor: Iteration failed
  #001: H5Gint.c line 842 in H5G_iterate(): error iterating over links
    major: Symbol table
    minor: Iteration failed
  #002: H5Gobj.c line 696 in H5G__obj_iterate(): can't iterate over compact links
    major: Symbol table
    minor: Iteration failed
  #003: H5Gcompact.c line 427 in H5G__compact_iterate(): iteration operator failed
    major: Symbol table
    minor: Can't move to next iterator location
  #004: H5Glink.c line 480 in H5G__link_iterate_table(): iteration operator failed
    major: Symbol table
    minor: Can't move to next iterator location
ncdump: /home/zender/foo.nc: NetCDF: HDF error

```

It is unclear whether this is related to #381, or any of the other rename issues documented [here](http://nco.sf.net/nco.html#ncrename_crd), but superficially it looks different to me (i.e., different HDF errors). Error occurs on all netCDF libraries tested, including 4.3.3.1, 4.4.1.1, and today's development snapshot.

**Comment by DennisHeimbigner** [comment]
can h5dump read it?

**Comment by czender** [comment]
Yes, output is too long to post so I put it [here](http://dust.ess.uci.edu/tmp/foo.h5dump).

**Comment by WardF** [comment]
Does it fail with previous versions of hdf5, or is that difficult to test? This will be fixed before our next release.

**Comment by WardF** [comment]
Hmm. I'll set up something to test previous versions, but I missed the part where h5dump can read it.  I'll also take a look.

**Comment by czender** [comment]
Yes it fails with older HDF too.

**Comment by DennisHeimbigner** [comment]
Can you also send us the h5dump of merra.nc (or send merra.nc) so we can
diff it with the above h5dump output?

**Comment by czender** [comment]
See the "test file" link in the OP

**Comment by WardF** [comment]
Ok, from gdb I've traced the failure (so far) to nc4file.c:2161.  The call to `H5Literate()` is returning an error (not currently captured, but I'll make the appropriate change to see what the return code is).

**Comment by WardF** [comment]
The diff from h5dump (for merra.nc4 and foo.nc)  is attached for reference.

* [diff.txt](https://github.com/Unidata/netcdf-c/files/1252591/diff.txt)

**Comment by WardF** [comment]
Ok @czender, I've taken a look at this and here's what I've found.

I wrote a C program that called `nc_rename_dim()` (attached below) and ran it against merra.nc4.  The program is pretty straight-forward, and changes the file in-place, but `ncdump` is able to read it and it looks as I would expect it to.  I ran h5dump against it and compared the output from *that* to the h5dump of the original unchanged file, and the differences are in line with what I'd expect to see.  At this point I wonder if the issue is somewhere in `ncrename`? Take a look at the example I've attached below and let me know if I'm missing a step or if there's a change I would need to make to more closely mimic what `ncrename` is doing.  For the most basic usage, however, `nc_rename_dim()` does not exhibit an obvious error.

* [rename_test.c.txt](https://github.com/Unidata/netcdf-c/files/1252642/rename_test.c.txt)
* [diff2.txt](https://github.com/Unidata/netcdf-c/files/1252644/diff2.txt)
